### PR TITLE
Episode 17.2 - Sage's Legacy

### PR DESCRIPTION
**Pull request by Everade**
* **Addressed Issue(s)**: 
https://github.com/rathena/rathena/issues/6470

* **Server Mode**: 
Renewal

* **Description of Pull Request**: 
Implementation of Episode 17.2 - Sage's Legacy (Walkthrough Conversion)

**ToDo:**
- [x] Reverse custom functions/cleanup
- [x] Uncomment start of 17.2 in the 17.1 quest line
- [x] Monsters (custom)
- [x] Monster Skills (custom)
- [x] Add missing barter shops/npc
- [ ] New mob skills implementation, `NPC_LOCKON_LASER`, `NPC_RELIEVE_ON`, `NPC_RELIEVE_OFF`, `NPC_RAINOFMETEOR`
- [x] Allow `unitskillusepos` script command to bypass the skill range restriction defined in `skill_db.yml`
- [ ] Independent Tester Feedback

Credits to @crazy-arashi and @JohnnyPlayy  for the initial script/content release

### CODE REVIEW

**Comment by CMSkull** [comment]
[Error]:
script error on npc/re/instances/HiddenGarden.txt line 1531
    parse_callfunc: not enough arguments, expected ','
  1526 :        end;
  1527 :
  1528 : OnTouch:
  1529 :        .@id = atoi(replacestr(strnpcinfo(2),"172_sec_warp_",""));
  1530 :        if (!getd("'warp_" + .@id) && .@id < 3) {
* 1531 :                setd("'warp_" + .@id = 1)';'
  1532 :                donpcevent instance_npcname("#172_sec_mob") + "::OnSummon";
  1533 :        }
  1534 :        switch (.@id) {
  1535 :                case 0: .@x = 101; .@y = 233; break;
  1536 :                case 1: .@x = 197; .@y = 200; break;
[Error]:
script error on npc/re/instances/LostFarm.txt line 809
    Unexpected newline at string.
   804 :        end;
   805 :
   806 : OnMobKill:
   807 :        if (!mobcount('map$,instance_npcname(strnpcinfo(0)) + "::OnMobKill"))
   808 :                'lost_farm",'lost_farm + 1);
*  809 :        end;
   810 : }
   811 :
   812 : 1@lost,38,296,0        duplicate(#172_lost_trap_0-0)   #172_lost_trap_0-1      HIDDEN_WARP_NPC,5,5
   813 : 1@lost,51,317,0        duplicate(#172_lost_trap_0-0)   #172_lost_trap_0-2      HIDDEN_WARP_NPC,5,5
   814 : 1@lost,92,317,0        duplicate(#172_lost_trap_0-0)   #172_lost_trap_0-3      HIDDEN_WARP_NPC,5,5

File: npc/re/instances/LostFarm.txt [hunk 1]
```diff
@@ -727,85 +720,94 @@ end;
...
-end;
+	for (.@i = 0; .@i < getarraysize(.@xy$); .@i += 4)
+		monster 'map$,atoi(.@xy$[.@i]),atoi(.@xy$[.@i+1]),.@xy$[.@i+3],atoi(.@xy$[.@i+2]),1,instance_npcname(strnpcinfo(0)) + "::OnMobKill";
+	end;
 
 OnMobKill:
-	if(!mobcount(get_instance_var("map$"),instance_npcname(strnpcinfo(0)) + "::OnMobKill"))
-		set_instance_var("lost_farm",get_instance_var("lost_farm") + 1);
-end;
+	if (!mobcount('map$,instance_npcname(strnpcinfo(0)) + "::OnMobKill"))
```
Comment (KSkull) [inline npc/re/instances/LostFarm.txt:808]: Unexpected newline at string.

**Comment by Megelith** [comment]
I feel like I'm missing something obvious here but I can't get the first 17.2 quest to initiate. The NPC Rookie just keeps repeating the dialogue from the end of 17.1. I have previously done all the 17.1 quests before merging this branch. The character I'm using has quest 16360 completed and does not have 18000 started.

**Comment by Kurtjcobain** [comment]
🤔
- Id: 100160
    AegisName: Auto_M_Box
    Name: Automatic Module Box
    **Type: Delayconsume**                    →     **Type: Usable**
    **Script: |**
      **laphine_synthesis();**        
                  ↓
    **Script: |**
   **getrandgroupitem(IG_AUTOMATIC_MODULE_MIX,1);**


  - Id: 100161
    AegisName: Epic_M_Box
    Name: Epic Module Box
    **Type: Delayconsume**                    →     **Type: Usable**
    **Script: |**
      **laphine_synthesis();**                    
                  ↓
    **Script: |**
   **getrandgroupitem(IG_EPIC_MODULE_MIX,1);**

**Comment by rainer8989** [comment]
Cant enter water garden for example, only Generate Instance can...always go to "else" menu, not to Enter dungeon

File: db/re/mob_db.yml [hunk 2]
```diff
@@ -92498,168 +92498,3081 @@ Body:
...
+      - Item: Light_Granule
+        Rate: 100
+      - Item: Scell
+        Rate: 1200
+      - Item: Red_Gemstone
+        Rate: 100
+      - Item: Plasma_R2_Card
+        Rate: 1
+        StealProtected: true
+  - Id: 20696
```
Comment (Pokye) [inline db/re/mob_db.yml:95461]: Information taken from the DP. Why Modes **Assist: True** ?
```c
  - Id: 20696
    AegisName: EP17_2_CHILD_ADMIN1
    Name: Child Admin Beta
    Level: 130
    Hp: 180
    Attack: 10
    Attack2: 10
    Defense: 10
    MagicDefense: 10
    Str: 10
    Agi: 10
    Vit: 10
    Int: 10
    Dex: 10
    Luk: 10
    AttackRange: 1
    SkillRange: 10
    ChaseRange: 12
    Size: Small
    Race: Formless
    Element: Neutral
    ElementLevel: 1
    WalkSpeed: 150
    AttackDelay: 1152
    AttackMotion: 1344
    DamageMotion: 180
    Ai: 01
  - Id: 20697
    AegisName: EP17_2_CHILD_ADMIN2
    Name: Child Admin Beta
    Level: 180
    Hp: 180
    Attack: 10
    Attack2: 10
    Defense: 10
    MagicDefense: 10
    Str: 10
    Agi: 10
    Vit: 10
    Int: 10
    Dex: 10
    Luk: 10
    AttackRange: 1
    SkillRange: 10
    ChaseRange: 12
    Size: Small
    Race: Formless
    Element: Neutral
    ElementLevel: 1
    WalkSpeed: 150
    AttackDelay: 1152
    AttackMotion: 1344
    DamageMotion: 480
    Ai: 01
    ```

[the remark this answers is not in the document]
Comment (Atemo) [reply db/re/mob_db.yml:95461]: Th data probably doesn't match as DP mobs were updated recently

**Comment by Dieby** [comment]
image

Does anyone know of this error and how to fix it?

**Comment by hannicaldummy** [comment]
There is a list with the mob skills missing?

**Review by null6385 — APPROVED** [review]

**Comment by ChingChuan-Chen** [comment]
> image
> 
> Does anyone know of this error and how to fix it?

This is related to https://github.com/rathena/rathena/issues/7114 , you can apply this PR to fix it, https://github.com/rathena/rathena/pull/7116 .

**Comment by Dieby** [comment]
image

It says that the instance is not created and I have everything added.... could someone solve this error?

File: db/re/quest_db.yml [hunk 3]
```diff
@@ -5560,9 +5598,40 @@ Body:
...
     TimeLimit: 4h
+  - Id: 8586
+    Title: Keeping Breeding Ground
   - Id: 8587
     Title: Daily farm experience - Standby
     TimeLimit: 4h
+  - Id: 8588
+    Title: Tail Collection
+  - Id: 8589
+    Title: It's Boss Meow!
```
Comment (Pokye) [inline db/re/quest_db.yml:5609]: Title: "It's Boss Meow!"

File: db/re/quest_db.yml [hunk 4]
```diff
@@ -10387,54 +10689,62 @@ Body:
...
+    Title: Conversation with Sweetie
   - Id: 18020
-    Title: Vitre's Songs
+    Title: Greeting
   - Id: 18021
-    Title: Vitre's Songs
+    Title: A guest
   - Id: 18022
-    Title: Vitre's Songs
+    Title: Hey!Sweetie
```
Comment (Pokye) [inline db/re/quest_db.yml:10737]: Title: Hey! Sweetie

File: db/re/quest_db.yml [hunk 5]
```diff
@@ -10387,54 +10689,62 @@ Body:
...
+    Targets:
+      - Mob: MD_SWEETY
+        Count: 1
   - Id: 18023
-    Title: Vitre the Spy
+    Title: Hey!Sweetie - Waiting
```
Comment (Pokye) [inline db/re/quest_db.yml:10742]: Title: Hey! Sweetie - Waiting

File: db/re/quest_db.yml (hunk 3 above)
Comment (Atemo) [reply db/re/quest_db.yml:5609]: (feel free to use the suggestion feature for small review)
Capture d’écran (1115)

**Comment by idk-whoami** [comment]
cbbead0da9522f367dcc3117de0b03366140c414
Capture

Remove the quest 11615 warp npc can't be used  :)


Capture

**Comment by idk-whoami** [comment]
Remove the quest 11617 warp npc can't be used

Capture



Remove the quest 11618 warp npc can't be used

Capture

**Comment by idk-whoami** [comment]
Capture

missing ! for check logical not
```
	if (!'water_garden) {
		mes "- You can enter the water garden after the party leader has entered the MD. -";
		close;
	}
```

**Comment by Dieby** [comment]
> [cbbead0](https://github.com/rathena/rathena/commit/cbbead0da9522f367dcc3117de0b03366140c414) Capture
> 
> Remove the quest 11615 warp npc can't be used :)
> 
> Capture

https://github.com/rathena/rathena/pull/6799#issuecomment-1324066853

which quest are used now I have the same error... :(

**Comment by Dieby** [comment]
image

File: npc/re/merchants/enchan_sage_legacy_17_2.txt [hunk 6]
```diff
@@ -285,10 +285,10 @@ ba_in01,87,373,3	script	Spiera#ep172_soapstone	MD_ASSISTANT,{
 
 //= Equipment Shop
 ba_in01,87,383,3	script	Yeoncheong#equipment_exchange	4_EP17_CLEANER_W,{
-	if (checkweight(1201,3) == 0) {	// todo
+	if (checkweight(1201,3) == 0) {
 		cutin "ep172_omega.bmp",2;
-		mes "[��û]";
-		mes "������ ���� á��ϴ�. ���� ���Կ� ���� ��ǰ ��� Ȯ�� �� �ٽ� ��� �ɾ� �ּ���.";
+		mes "[ Yeoncheong ]";
```
Comment (Lemongrass3110) [inline npc/re/merchants/enchan_sage_legacy_17_2.txt:290]: Original text did not have blanks around the name.

File: db/re/mob_db.yml [hunk 7]
```diff
@@ -95916,168 +95916,3405 @@ Body:
...
+    Size: Small
+    Race: Formless
+    Element: Neutral
+    ElementLevel: 1
+    WalkSpeed: 150
+    AttackDelay: 1152
+    AttackMotion: 1344
+    DamageMotion: 480
+    Ai: 01
+    Drops:
```
Comment (Pokye) [inline db/re/mob_db.yml:99236]:
```suggestion
```

File: db/re/mob_db.yml [hunk 8]
```diff
@@ -95916,168 +95916,3405 @@ Body:
...
```
Comment (Pokye) [inline db/re/mob_db.yml:99208]:
```suggestion
```

**Review by Pokye — COMMENTED** [review]
Remove as it is not being used.

**Comment by Atemo** [comment]
If anyone knows how `NPC_LOCKON_LASER`, `NPC_RELIEVE_ON`, `NPC_RELIEVE_OFF`, `NPC_RAINOFMETEOR` skills work, please feel free to open a PR.

**Comment by eppc0330** [comment]
Nice

**Comment by Dieby** [comment]
image

When I finish applying everything new I have this error, does anyone have the solution?

**Comment by fab10Carvalho** [comment]
> image
> 
> When I finish applying everything new I have this error, does anyone have the solution?

may be https://github.com/rathena/rathena/pull/7591 But im not sure :D

**Comment by Lemongrass3110** [comment]
@fab10Carvalho you are correct. This is needed for it.

**Comment by Dieby** [comment]
> > image
> > When I finish applying everything new I have this error, does anyone have the solution?
> 
> may be #7591 But im not sure :D

thank you very much it is solved

**Comment by Dieby** [comment]
I was testing the latest rathena emulator and 17.2 in the water garden instance in phase 3, the mobs do not appear in normal or hard, I already tried reinstalling everything and nothing

image

**Pull request merged**